Update the development docs (#365)
looker/actions: docs/action_api.md

# Looker Action API

To build an action directly into an existing web service in any language, you can have your server implement our Action API directly.

The Action API is a simple webhook-like API to accept actions from Looker. Your server will provide a few endpoints that Looker can call to list and execute actions.

By implementing the Action API your server becomes an "Action Hub" that can be connected to one or more Looker instances. An Action Hub can expose one or more actions.

## Endpoints

> In order to implement the Action API your server will need to expose a few endpoints for Looker to call:

Reviewer: ```suggestion
To implement the Action API, your server must expose a few endpoints for Looker to call:
```

- [Actions List Endpoint](#actions-list-endpoint)

  This endpoint lists all the actions your hub exposes. Each listed action contains metadata instructing Looker how to execute your action. It indicates what data payload formats are supported, what information (if any) to collect from the Looker administrator (such as API keys), and if the action also has an end-user facing Action Form.

>   The URL of this endpoint is entered into the Looker admin panel (as the "Action Hub URL") and is how admins connect to your action hub.

Reviewer: ```suggestion
  The URL of this endpoint is entered into the Looker admin panel (as the **Action Hub URL**) and is how admins connect Looker to your action hub.
```

- [Action Execute Endpoint](#action-execute-endpoint)

>   This is the endpoint to which Looker will send the data payload, form information, and other metadata in order to execute a given action. This is the main implementation of your action. This endpoint works the same way as a Looker webhook endpoint.

Reviewer: ```suggestion
  This is the endpoint to which Looker sends the data payload, form information, and other metadata to execute a given action. This is the main implementation of your action. This endpoint works the same way as a Looker webhook endpoint.
```

- (optional) [Action Form Endpoint](#action-form-endpoint)

>   If the action's definition specifies a form, Looker will ask this endpoint for a form template that will be displayed to the end user to configure the action (e.g. when using the action as a destination in the Schedule dialog). Since Looker will query this endpoint every time the form is displayed, it's possible to dynamically populate the form based on whatever information you like. For example, for a chat application the form might have a dropdown that dynamically list all the chat channels available at that moment. Or for an action that uses OAuth with the Google Analytics API, the form could dynamically list the web properties to which the current user has access.

Reviewer: ```suggestion
  If the action's definition specifies a form, Looker asks this endpoint for a form template to display to the user to configure the action (e.g. when using the action as a destination in the Schedule dialog). Since Looker will query this endpoint every time the form is displayed, it's possible to dynamically populate the form with custom user input fields. For example, for a chat application the form might have a drop-down that dynamically lists all the chat channels available at that moment. Or for an action that uses OAuth with the Google Analytics API, the form could dynamically list the web properties to which the current user has access.
```

```ts
interface ActionList {
  /** A label for the Hub that will be shown in the Looker admin page **/
  label: string
  /** (required) An array of action definition objects. */
  integrations: ActionDefinition[]
}

interface ActionDefinition {
  /** (required) A unique name for the action. This should be unique across all actions in the hub. */
  name: string
  /** (required) A human-readable label for the action. */
  label: string
  /** Description of the action. */
  description?: string
  /** A Data URI reprensenting an icon image for the action. */
  icon_data_uri?: string
  /** (required) An absolute URL of the execute endpoint for this action. */
  url: string
  /** An absolute URL of the form endpoint for this action. */
  form_url?: string
  /** Array of params for the action. */
  params?: Param[]

>   /** A list of descriptions of required fields that this action is compatible with. If there are multiple entries in this list, the action requires more than one field. */

Reviewer: Not sure if you want to add the `user_attribute_params` info here? https://docs.looker.com/sharing-and-publishing/action-hub#adding_user_attributes_to_custom_actions

Author: good point - done!

### Action User State

> The action hub API is designed to keep each action independent from previous actions. For certain actions, however - especially OAuth enabled actions - state may be required per user. If there is state for a user and action then Looker will send the state as part of a `form` or `execute` request. If there is no state then the field is unused. If you wish to set state that can be accomplished by providing a `state` property in an `execute` or `form` response. A refresh time may also be specified in seconds, which causes Looker to re-query the `form` endpoint for that user after the refresh time has elapsed (with a current maximum of once every 10 minutes). This can be used to keep authentication credentials up-to-date or to simply store some state for the user & action combo within the Looker server.

Reviewer: ```suggestion
The action hub API is designed to keep each action independent from other actions. For certain actions, however - especially OAuth-enabled actions - state may be required per user. If there is state for a user and action, Looker sends the state as part of a `form` or `execute` request. If there is no state, the field is unused. If you wish to set state, you can provide a `state` property in an `execute` or `form` response. A refresh time may also be specified in seconds, which causes Looker to re-query the `form` endpoint for that user after the refresh time has elapsed (with a current maximum of once every 10 minutes). This can be used to keep authentication credentials up-to-date or to simply store some state for the user-and-action combination within the Looker server.
```
